src/fhirtypes/StructureDefinition.ts

```diff
@@ -200,6 +253,95 @@ export class StructureDefinition {
     }
     return sd;
   }
+
+  /**
+   * Parses a FSH Path into a more easily usable form
+   * @param {string} fshPath - A syntactically valid path in FSH
+   * @returns {PathPart[]} an array of PathParts that is easier to work with
+   */
+  private parseFSHPath(fshPath: string): PathPart[] {
+    const pathParts: PathPart[] = [];
+    const splitPath = fshPath.split('.');
+    for (const pathPart of splitPath) {
+      const splitPathPart = pathPart.split('[');
+      if (splitPathPart.length === 1 || pathPart.endsWith('[x]')) {
+        // There are no brackets, or the brackets are for a choice, so just push on the name
+        pathParts.push({ base: pathPart });
+      } else {
+        // We have brackets, let's  save the bracket info
+        const fhirPathBase = splitPathPart[0];
+        // Get the bracket elements and slice off the trailing ']'
+        const brackets = splitPathPart.slice(1).map(s => s.slice(0, -1));
+        pathParts.push({ base: fhirPathBase, brackets: brackets });
+      }
+    }
+    return pathParts;
+  }
+
+  /**
+   * Looks for a matching choice element and if found slices it, adds the slices to the structdef
+   * and then returns the newly created slice.
+   * @param {string} fhirPath - The path in FHIR to match with
+   * @param {ElementDefinition[]} elements - The set of elements to search through
+   * @returns {ElementDefinition} - The new slice element if found, else undefined
+   */
+  private sliceMatchingValueX(fhirPath: string, elements: ElementDefinition[]): ElementDefinition {
+    let matchingType: ElementDefinitionType;
+    const matchingXElement = elements.find(e => {
+      if (e.path.endsWith('[x]')) {
+        for (const t of e.type) {
+          if (`${e.path.slice(0, -3)}${capitalize(t.code)}` === fhirPath) {
+            matchingType = t;
+            return true;
+          }
+        }
+      }
+    });
+    if (matchingXElement) {
+      // If we find a matching [x] element, we need to slice it to create the child element
+      matchingXElement.sliceIt('type', '$this', false, 'open');
+      // Get the sliceName for the new element
+      const sliceName = fhirPath.slice(fhirPath.lastIndexOf('.') + 1);
+      const newSlice = matchingXElement.addSlice(sliceName, matchingType);
+      return newSlice;
+    }
+    return;
+  }
+
+  /**
+   * Looks for a slice within the set of elements that matches the fhirPath
+   * @param {PathPart} pathPart - The path to match sliceName against
+   * @param {ElementDefinition[]} elements - The set of elements to search through
+   * @returns {ElementDefinition} - The sliceElement if found, else undefined
+   */
+  private findMatchingSlice(pathPart: PathPart, elements: ElementDefinition[]): ElementDefinition {
+    return elements.find(e => e.sliceName === pathPart.brackets.join('/'));
```

Comment:
We talked before about a possibility of `foo[sliceName][referenceName]` -- will this erroneously try to find a slice called `sliceName/referenceName`?

Reply:
It will try to find `sliceName/referenceName`, but that should fail so this will return null, and then we will go on to the `findMatchingRef` function, which will then treat the last bracketed element as a reference, so `referenceName` here, and search for the element with a slice `sliceName` that has a reference to `referenceName`. If there does exist a re-slice with the name `sliceName/referenceName`, and a slice `sliceName` with a reference to `referenceName`, we wouldn't have a way to distinguish between those two, but I don't think there really is a way. Plus that seems very improbable.

Reply:
Ah.  OK.  I think it's cleaner as separate functions like this, but it's a little unfortunate that this function kind of depends on another function to be called after it in order to really get the right answer. If we re-use this function anywhere else (which is probably doubtful?), that's something we'll need to keep in mind.  Maybe we can at least document that behavior in the comment?

Reply:
Yeah I can document the ambiguity involved.